src/table/tests.rs

```diff
@@ -344,3 +370,67 @@ fn test_table_big_values() {
     // TODO: support max_version in table
     // assert_eq!(n, table.max_version());
 }
+
+#[test]
+fn test_iterator_error_eof() {
+    let opts = get_test_table_options();
+    let table = build_test_table(b"key", 10000, opts);
+
+    let mut it = table.new_iterator(0);
+    it.rewind();
+
+    while it.valid() {
+        it.next();
+    }
+
+    assert!(matches!(it.error(), Some(IteratorError::EOF)));
+}
+
+#[test]
+fn test_iterator_use_without_init() {
+    let opts = get_test_table_options();
+    let table = build_test_table(b"key", 1000, opts);
+    let mut it = table.new_iterator(0);
+    // Generally, developers should call `rewind` before using an iterator.
+    // If iterator is not initialized, getting key directly from iterator
+    // will cause panic. Directly calling `next` will return
+    // the first entry.
+    it.next();
+    assert_eq!(user_key(it.key()), key(b"key", 0));
+}
+
+#[test]
+fn test_iterator_out_of_bound() {
+    let opts = get_test_table_options();
+    let table = build_test_table(b"key", 1000, opts);
+    let mut it = table.new_iterator(0);
+    it.seek_to_last();
+    assert!(it.error().is_none());
+    it.next();
+    assert!(matches!(it.error(), Some(IteratorError::EOF)));
+    it.next();
+    assert!(matches!(it.error(), Some(IteratorError::EOF)));
+    it.next();
+    assert!(matches!(it.error(), Some(IteratorError::EOF)));
+    it.rewind();
+    assert!(it.error().is_none());
+    assert_eq!(user_key(it.key()), key(b"key", 0));
+}
+
+#[test]
+fn test_iterator_out_of_bound_reverse() {
+    let opts = get_test_table_options();
+    let table = build_test_table(b"key", 1000, opts);
+    let mut it = table.new_iterator(ITERATOR_REVERSED);
+    it.seek_to_first();
+    assert!(it.error().is_none());
+    it.next();
+    assert!(matches!(it.error(), Some(IteratorError::EOF)));
```

Comment:
Can `IteratorError` derive `PartialEq`? So you can use `assert_eq`, which produces more readable error when fail.

Reply:
I've added `PartialEq`, and now in tests, we use `== Some` instead of `matches!`. But in places other than test, I retained the original implementation (`matches!(...)`). Do we need to also change how it works in main program?